pandas/_libs/missing.pyx

```diff
@@ -0,0 +1,176 @@
+# -*- coding: utf-8 -*-
+# cython: profile=False
+
+from cpython cimport PyFloat_Check, PyComplex_Check
+
+cimport cython
+from cython cimport Py_ssize_t
+
+import numpy as np
+cimport numpy as np
+from numpy cimport ndarray, int64_t, uint8_t
+np.import_array()
+
+cimport util
+
+from tslibs.np_datetime cimport get_timedelta64_value, get_datetime64_value
+from tslibs.nattype import NaT, iNaT
+
+cdef double INF = <double> np.inf
+cdef double NEGINF = -INF
+
+cdef int64_t NPY_NAT = util.get_nat()
+
+
+cdef inline bint is_null_datetimelike(v):
+    # determine if we have a null for a timedelta/datetime (or integer
+    # versions)
+    if util._checknull(v):
+        return True
+    elif v is NaT:
+        return True
+    elif util.is_timedelta64_object(v):
+        return v.view('int64') == iNaT
+    elif util.is_datetime64_object(v):
+        return v.view('int64') == iNaT
+    elif util.is_integer_object(v):
+        return v == iNaT
+    return False
+
+
+cdef inline bint _check_all_nulls(object val):
+    """ utility to check if a value is any type of null """
+    cdef bint res
+    if PyFloat_Check(val) or PyComplex_Check(val):
+        res = val != val
+    elif val is NaT:
+        res = 1
+    elif val is None:
+        res = 1
+    elif util.is_datetime64_object(val):
+        res = get_datetime64_value(val) == NPY_NAT
+    elif util.is_timedelta64_object(val):
+        res = get_timedelta64_value(val) == NPY_NAT
+    else:
+        res = 0
+    return res
+
+
+cpdef bint checknull(object val):
+    if util.is_float_object(val) or util.is_complex_object(val):
+        return val != val  # and val != INF and val != NEGINF
+    elif util.is_datetime64_object(val):
+        return get_datetime64_value(val) == NPY_NAT
+    elif val is NaT:
+        return True
+    elif util.is_timedelta64_object(val):
+        return get_timedelta64_value(val) == NPY_NAT
+    elif util.is_array(val):
+        return False
+    else:
```

Comment:
is there a reason for not pulling in ``util._checknull`` here as well, as seems logical? (or just future PR)

Reply:
You mean defining it here instead of in `util`?  Or importing it into the namespace?  I'd be +1 on the former, indifferent to the latter.

Reply:
I think for sure *should* define it here. but then this puts missing as a dep of things like hashing.pyx. ok with it being a dep of any of the tslibs though.

```
(pandas) bash-3.2$ find pandas -name '*.pyx' | xargs grep _checknull
pandas/_libs/hashing.pyx:from util cimport _checknull
pandas/_libs/hashing.pyx:        elif _checknull(val):
pandas/_libs/interval.pyx:        if util._checknull(interval):
pandas/_libs/lib.pyx:from util cimport is_array, _checknull, _checknan
pandas/_libs/lib.pyx:        return _checknull(val)
pandas/_libs/lib.pyx:        return _checknull(val)
pandas/_libs/lib.pyx:        result[i] = val is NaT or util._checknull_old(val)
pandas/_libs/lib.pyx:                _checknull(x) and _checknull(y)):
pandas/_libs/lib.pyx:    if _checknull(val):
pandas/_libs/lib.pyx:        if _checknull(x):
pandas/_libs/lib.pyx:            if _checknull(x):
pandas/_libs/lib.pyx:            elif _checknull(y):
pandas/_libs/src/inference.pyx:        if util._checknull(val):
pandas/_libs/src/inference.pyx:        elif util._checknull(v):
pandas/_libs/src/inference.pyx:    if util._checknull(v):
pandas/_libs/src/inference.pyx:    if util._checknull(v):
pandas/_libs/src/inference.pyx:    if util._checknull(v):
pandas/_libs/src/inference.pyx:    if util._checknull(v):
pandas/_libs/src/inference.pyx:        return util._checknull(value)
pandas/_libs/src/inference.pyx:            bint is_generic_null = util._checknull(value)
pandas/_libs/tslib.pyx:from tslibs.nattype cimport _checknull_with_nat, NPY_NAT
pandas/_libs/tslib.pyx:        if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:        if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslib.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslibs/nattype.pyx:cdef inline bint _checknull_with_nat(object val):
pandas/_libs/tslibs/strptime.pyx:from nattype cimport _checknull_with_nat, NPY_NAT
pandas/_libs/tslibs/strptime.pyx:            if _checknull_with_nat(val):
pandas/_libs/tslibs/timedeltas.pyx:from nattype cimport _checknull_with_nat, NPY_NAT
pandas/_libs/tslibs/timedeltas.pyx:    if _checknull_with_nat(ts):
pandas/_libs/tslibs/timedeltas.pyx:    if _checknull_with_nat(other):
pandas/_libs/tslibs/timedeltas.pyx:        elif _checknull_with_nat(value):
```

Reply:
If we're moving util._checknull anyway, I'd advocate renaming it to e.g. `check_none_or_nan`

Reply:
sure

Reply:
> I think for sure should define it here. but then this puts missing as a dep of things like hashing.pyx. ok with it being a dep of any of the tslibs though.

I'll take a look and see which util funcs can be moved without messing with dependencies.

FWIW this PR already adds missing to the 'pxdfiles` key of hashtable, which cimports missing.checknull.  Previously it was an un-declared dependency on lib.

Reply:
Looks like util._checknull_old can be moved to missing (is used there once, nowhere else).  Let's saving util._checknull for later, since it is used in a bunch of places.

Reply:
ok with leaving these for later as well

Reply:
Sounds good.  Just pushed commit with docstrings.